controllers/nstemplatetierrevisioncleanup/nstemplatetier_revision_cleanup_controller.go

```diff
@@ -0,0 +1,128 @@
+package nstemplatetierrevisioncleanup
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	"github.com/codeready-toolchain/host-operator/controllers/nstemplatetier"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const deletionTimeThreshold = 30 * time.Second
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *Reconciler) SetupWithManager(mgr manager.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&toolchainv1alpha1.TierTemplateRevision{}).
+		Complete(r)
+}
+
+type Reconciler struct {
+	Client runtimeclient.Client
+	Scheme *runtime.Scheme
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// fetch the NSTemplateTier tier
+	ttr := &toolchainv1alpha1.TierTemplateRevision{}
+	if err := r.Client.Get(ctx, request.NamespacedName, ttr); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("TierTemplateRevision not found")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		logger.Error(err, "unable to get the current TierTemplateRevision")
+		return reconcile.Result{}, fmt.Errorf("unable to get the current TierTemplateRevision: %w", err)
+	}
+	tierName, ok := ttr.GetLabels()[toolchainv1alpha1.TierLabelKey]
+	if !ok {
+		return reconcile.Result{}, fmt.Errorf("tier-name label not found in tiertemplaterevision")
+
+	}
+	// fetch the related NSTemplateTier tier
+	tier := &toolchainv1alpha1.NSTemplateTier{}
+	if err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: ttr.GetNamespace(),
+		Name:      tierName,
+	}, tier); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("NSTemplateTier not found")
+			return reconcile.Result{}, nil
+		}
```

Comment:
We cannot return here, if the NSTemplateTier is removed, then we should eventually remove the TTR CRs as well